File: experiments/run_first_sim.py
# -*- coding: utf-8 -*-
"""首次完整仿真：Python -> MATLAB -> Simulink -> 读结果"""
import subprocess, time, os, sys

PROJECT = 'F:/CLAUDE/research/ems-platform'
RESULTS = os.path.join(PROJECT, 'results')
os.makedirs(RESULTS, exist_ok=True)

print('=' * 50)
print('第1次完整仿真')
print(f'模型: Energy.slx')
print(f'MATLAB: F:/Matlab/bin/matlab.exe')
print('=' * 50)

t0 = time.time()
proc = subprocess.Popen(
    ['F:/Matlab/bin/matlab.exe', '-batch',
     "run('F:/CLAUDE/research/ems-platform/experiments/run_first_sim.m')"],
    stdout=subprocess.PIPE, stderr=subprocess.PIPE
)

try:
    stdout, stderr = proc.communicate(timeout=300)
    out = stdout.decode('cp936', errors='replace')
    err = stderr.decode('cp936', errors='replace')

    print(f'\n⏱ 总耗时: {time.time()-t0:.1f}s (退出码: {proc.returncode})')
    print('-' * 50)

    # 显示关键输出
    for line in out.split('\n'):
        line_s = line.strip()
        if line_s and ('工作区' in line_s or '找到' in line_s or '开始' in line_s
                       or '完成' in line_s or '结果' in line_s or '耗时' in line_s
                       or '变量' in line_s or '模型' in line_s):
            print(f'  {line_s}')

    if err.strip():
        print(f'\nMATLAB 错误输出: {err[:500]}')

    # 检查结果文件
    mat_file = os.path.join(RESULTS, 'sim_first_run.mat')
    if os.path.exists(mat_file):
        size = os.path.getsize(mat_file)
        print(f'\n✅ 仿真结果已保存: {mat_file} ({size/1024:.1f}KB)')
    else:
        print('\n⚠️  未生成结果文件（首次空跑正常）')

except subprocess.TimeoutExpired:
    proc.kill()
    print('❌ MATLAB 超时（300s）')
except Exception as e:
    print(f'❌ 错误: {e}')

File: experiments/run_first_sim.m
% run_first_sim.m
% 第1次完整仿真：加载Energy.slx模型，跑仿真，导出结果
% 被Python调用：matlab -batch run('...run_first_sim.m')

project_dir = 'F:/CLAUDE/research/ems-platform';
model_dir = fullfile(project_dir, 'env/simulink_models');
results_dir = fullfile(project_dir, 'results');

% 切换到模型目录
cd(model_dir);

% 加载工作区数据（模型需要这些变量）
if exist('matlab.mat', 'file')
    load('matlab.mat');
    fprintf('工作区数据已加载\n');
end

% 加载模型
load_system('Energy.slx');
fprintf('模型已加载: Energy.slx\n');

% 查看模型中的ToWorkspace模块
ws_blocks = find_system('Energy', 'BlockType', 'ToWorkspace');
fprintf('找到 %d 个ToWorkspace模块\n', length(ws_blocks));
for i = 1:length(ws_blocks)
    var_name = get_param(ws_blocks{i}, 'VariableName');
    fprintf('  [%d] %s -> 变量: %s\n', i, ws_blocks{i}, var_name);
end

% 查看Scope模块
scopes = find_system('Energy', 'BlockType', 'Scope');
fprintf('找到 %d 个Scope模块\n', length(scopes));

% 跑仿真
fprintf('\n开始仿真...\n');
tic;
sim('Energy.slx');
elapsed = toc;
fprintf('仿真完成! 耗时 %.2f秒\n', elapsed);

% 检查有哪些变量被写入了工作区
vars = who;
fprintf('\n工作区变量 (%d个):\n', length(vars));
for i = 1:length(vars)
    v = vars{i};
    if evalin('base', ['isa(', v, ', ''struct'') || isnumeric(', v, ')'])
        s = evalin('base', ['size(', v, ')']);
        fprintf('  %s (%dx%d)\n', v, s(1), s(2));
    end
end

% 保存关键数据到CSV用于Python读取
% （先做一次空跑看看哪些变量生成，再决定导出什么）
save(fullfile(results_dir, 'sim_first_run.mat'));

fprintf('\n结果已保存到: %s\n', fullfile(results_dir, 'sim_first_run.mat'));
fprintf('首次仿真完成!\n');
quit;
